research/conditional/moe_layers/token_choice.py

```diff
@@ -124,20 +139,38 @@ def choose_expert(self, gate_out) -> tuple[torch.Tensor, torch.Tensor]:
         if checkpointing_enabled:
             if checkpointing.is_in_first_forward():
                 with torch.no_grad():
-                    expert_index = torch.argmax(gate_out, dim=1, keepdim=True)
+                    _, expert_index = torch.topk(
+                        gate_out, k=self.experts_per_token, dim=1
+                    )
                     self._checkpointed_expert_index = expert_index
 
             if checkpointing.is_in_second_forward():
                 with torch.no_grad():
                     expert_index = self._checkpointed_expert_index
-                    assert isinstance(expert_index, torch.Tensor)
 
-            expert_gate = torch.gather(gate_out, dim=1, index=expert_index).squeeze()
-            expert_index = expert_index.squeeze()
+            assert isinstance(expert_index, torch.Tensor)
+            expert_gate = torch.gather(gate_out, dim=1, index=expert_index)
         else:
-            expert_gate, expert_index = torch.max(gate_out, dim=1)
+            expert_gate, expert_index = torch.topk(
+                gate_out, k=self.experts_per_token, dim=1
+            )
         return expert_gate, expert_index
 
+    def log_heavy(self):
+        return {
+            "gradient_of_gate_distribution": make_histogram(
+                self.gate.grad.flatten()
+            ),  # move
+            "gate_softmax_all_values": make_histogram(
+                self.logging_cache["gate_softmax_all_values"].flatten()  # move
+            ),
+            "tokens_per_expert_counts": make_histogram(
+                self.logging_cache["tokens_per_expert"]
+            ),
+            "load_balancing_loss": self.logging_cache["load_balancing_loss"],
+            "dropped_tokens_ratio": self.logging_cache["dropped_tokens_ratio"],
+        }
```

Comment:
does this work? I've had problems before. I'd love to see a neptune run with `logging_interval_heavy` set to like 10 or sth.

Reply:
Next PR